pallets/migrations/src/lib.rs

```diff
@@ -44,6 +44,18 @@ pub trait Migration {
 	/// constraints will lead to a bricked chain upon a runtime upgrade because the parachain will
 	/// not be able to produce a block that the relay chain will accept.
 	fn migrate(&self, available_weight: Weight) -> Weight;
+
+	/// Run a standard pre-runtime test. This works the same way as in a normal runtime upgrade.
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade(&self) -> Result<(), &'static str> {
+		Ok(())
+	}
+
+	/// Run a standard post-runtime test. This works the same way as in a normal runtime upgrade.
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade(&self) -> Result<(), &'static str> {
+		Ok(())
+	}
```

Comment:
At this point you're basically re-inventing https://crates.parity.io/frame_support/traits/trait.OnRuntimeUpgrade.html

What do you think about this instead:
```rust
pub trait Migration: OnRuntimeUpgrade {
  fn friendly_name(&self) -> &str;
}
```

This means you can only implement `Migration` with on a type that already implements `OnRuntimeUpgrade`. So there is no need to duplicate the methods. Currently you `migrate` method has a slightly different signature than `on_runtime_upgrade` so we would have to change one to match the other, or keep them separate and provide an implementation for your `migrate` method that calls the underlying `on_runtime_upgrade` method.

Reply:
I think there's a fundamental problem we would need to solve to do something like this and it would require some larger redesign. 

The definition `type MigrationsList: Get<Vec<Box<dyn Migration>>>;` is the problem. I believe the `dyn Migration` part requires a `vtable` and this can only be satisfied if every trait fn takes a `&self` parameter. (I believe that's the basic idea; I'd love for someone to clarify if I'm not entirely correct about that).

I also have a minor concern about reusing the name `on_runtime_upgrade` for the fn name because it might lead to confusion. Substrate will automatically invoke `on_runtime_upgrade`, which could be a pretty catastrophic thing to mix-up, although it would probably be hard to do since we're talking about a trait vs. a hook impl.

Reply:
Oh, good point about the `vtable`. I didn't think about that. I'm fine merging this with the current design.

I think if we do want to explore this route, the solution this problem would be to switch from using a `Vec` and dynamic dispatch, and instead use https://github.com/bkchr/impl-trait-for-tuples. This is precedented in Substrate already.

Reply:
Sounds good to me. I'd like to explore that, if only to learn about it. I've leveled up a bit in rust since writing this pallet, so it would be a useful exercise to review my designs at some point.

Reply:
I made a followup ticket for this https://purestake.atlassian.net/browse/MOON-932